canister/src/validation.rs

```diff
@@ -16,20 +16,47 @@ impl<'a> ValidationContext<'a> {
         let prev_block_hash = header.prev_blockhash.into();
         let chain = unstable_blocks::get_chain_with_tip(&state.unstable_blocks, &prev_block_hash)
             .ok_or_else(|| BlockDoesNotExtendTree(header.block_hash().into()))?
-            .into_chain();
+            .into_chain()
+            .iter()
+            .map(|block| (block.header(), block.block_hash()))
+            .collect();
 
         Ok(Self { state, chain })
     }
+
+    fn insert_in_chain(&mut self, item: (&'a BlockHeader, crate::types::BlockHash)) {
+        self.chain.push(item)
+    }
+    /// Initialize a `ValidationContext` for the given block header.
+    /// The given block header can be in the 'NextBlockHeaders'.
+    pub fn new_with_next_block_headers(
```

Comment:
This method just creates ValidationContex with a chain field extended for a blocks in next_block_headers.

Reply:
Is there a reason to not be extending the chain with the next block headers? i.e. why not use this contructor everytime and delete the the `new` method?

Reply:
Counterexample for not using the `new_with_next_block_headers` everywhere:
If we have two blocks `A` and `B`, and the previous block of `B` is `A`, and we receive only `A`'s header in the next block headers followed by receiving complete block `B`. If we use this method, validation will pass because it will find `B`'s predecessor(`A`) in `chain`, but it should fail because we still did not receive the whole block `A`.